sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -0,0 +1,275 @@
+        ///
+        /// <returns>The AMQP message that represents a batch containing the converted set of event data.</returns>
+        ///
+        public virtual AmqpMessage CreateBatchFromMessages(IEnumerable<AmqpMessage> source,
+                                                           string partitionKey = null)
+        {
+            Guard.ArgumentNotNull(nameof(source), source);
+            return BuildAmqpBatchFromMessages(source, partitionKey);
+        }
+
+        /// <summary>
+        ///   Builds a batch <see cref="AmqpMessage" /> from a set of <see cref="EventData" />
+        ///   optionally propagating the custom properties.
+        /// </summary>
+        ///
+        /// <param name="source">The set of events to use as the body of the batch message.</param>
+        /// <param name="copyCustomProperties"><c>true</c> if the custom properties of the event should be propagated to the message; otherwise, <c>false</c>.</param>
+        /// <param name="partitionKey">The partition key to annotate the AMQP message with; if no partition key is specified, the annotation will not be made.</param>
+        ///
+        /// <returns>The batch <see cref="AmqpMessage" /> containing the source events.</returns>
+        ///
+        private static AmqpMessage BuildAmqpBatchFromEvents(IEnumerable<EventData> source,
+                                                            bool copyCustomProperties,
+                                                            string partitionKey) =>
+            BuildAmqpBatchFromMessages(
+                source.Select(eventData => BuildAmqpMessageFromEvent(eventData, copyCustomProperties, partitionKey)),
+                partitionKey);
+
+        /// <summary>
+        ///   Builds a batch <see cref="AmqpMessage" /> from a set of <see cref="AmqpMessage" />.
+        /// </summary>
+        ///
+        /// <param name="source">The set of messages to use as the body of the batch message.</param>
+        /// <param name="partitionKey">The partition key to annotate the AMQP message with; if no partition key is specified, the annotation will not be made.</param>
+        ///
+        /// <returns>The batch <see cref="AmqpMessage" /> containing the source messages.</returns>
+        ///
+        private static AmqpMessage BuildAmqpBatchFromMessages(IEnumerable<AmqpMessage> source,
+                                                              string partitionKey)
+        {
+            AmqpMessage batchEnvelope;
+
+            var batchMessages = source.ToList();
+
+            if (batchMessages.Count == 1)
+            {
+                batchEnvelope = batchMessages[0];
+            }
+            else
+            {
+                batchEnvelope = AmqpMessage.Create(batchMessages.Select(message =>
+                {
+                    message.Batchable = true;
+                    using var messageStream = message.ToStream();
```

Comment:
Wow. I haven't seen this using construct. Must be a new language feature!

Reply:
Yup.   C# 8, but we're using a preview library to enable these.  :)